File: sdks/java/src/main/java/com/vivid/sdk/Subscription.java
package com.vivid.sdk;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;

/**
 * A handle to a subscription.
 * <p>
 * A subscription can be cancelled to stop receiving updates.
 */
public interface Subscription {
    Empty EMPTY = new Empty();

    static Subscription composite(Subscription... layers) {
        return composite(Arrays.asList(layers));
    }

    static Subscription composite(List<Subscription> layers) {
        if (layers.isEmpty()) {
            return EMPTY;
        }
        return new CompositeSubscription(layers);
    }

    /**
     * Cancel the subscription.
     * <p>
     * Once cancelled, no more updates will be received.
     */
    void cancel();

    class CompositeSubscription implements Subscription {
        private final Collection<Subscription> layers;

        public CompositeSubscription(Collection<Subscription> layers) {
            this.layers = layers;
        }

        @Override
        public void cancel() {
            layers.forEach(Subscription::cancel);
        }
    }

    class Empty implements Subscription {
        @Override
        public void cancel() {
            // NoOp
        }
    }

}

File: sdks/kotlin/src/main/kotlin/com/vivid/sdk/Subscription.kt
package com.vivid.sdk

/**
 * A handle to a subscription.
 *
 * A subscription can be cancelled to stop receiving updates.
 */
interface Subscription {

    /**
     * Cancel the subscription.
     *
     * Once cancelled, no more updates will be received.
     */
    fun cancel()

    companion object {
        operator fun invoke(layers: Collection<Subscription>) = CompositeSubscription(layers)
    }

    class CompositeSubscription(private val layers: Collection<Subscription>) : Subscription {
        override fun cancel() {
            layers.forEach { it.cancel() }
            if (layers is MutableCollection<*>) {
                layers.clear()
            }
        }
    }

    object Empty : Subscription {
        override fun cancel() {}
    }
}
